Fix time shift metric lookup

flash1293 opened the pull request:
Fixes https://github.com/elastic/kibana/issues/119767

When looking up the agg instance for a key in the response object during merging of the time split to check which time shift a metric belongs to, it would do a prefix check instead of a full check to compare the key with the id of the agg. In most cases this works fine, but if the ids of the metrics are a prefix of another agg (like `"1"` vs `"10"`), then the lookup can return the wrong metric and thus map it back to the wrong time shift.

This PR makes sure the id fully matches and adds a functional test for this case.

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

elasticmachine commented:
Pinging @elastic/kibana-app-services (Team:AppServicesSv)

mbondyra commented:
@elasticmachine merge upstream

mbondyra reviewed (approved):
Tested and the issue is fixed. Code LGTM 👌🏼

Dosant reviewed (approved):
Fix LGTM, didn't test

flash1293 commented:
OK, so I adjusted the fix a bit to make it more robust (also for the future) - unfortunately this made it more complex as well.

To make the tests better at detecting problems, I added a check whether there is actually data in the table - before it would just check whether the metric columns are shifted correctly respective to each other, but `null` is always matching `null` so it would pass the check.

To look up the correct agg config from the object key of the ES response object, it introduced a new method `getResponseId` on the agg config which allows you to do exactly this. In case of filtered metric, it's overwritten by the id of the nested "filter" bucket - this is necessary because the filtered metric agg is special in the sense that it doesn't return a nested object in the response.

flash1293 commented:
@mbondyra @Dosant could you have another look once it turns green?

mbondyra commented:
Reapproved!

Dosant commented on src/plugins/data/common/search/aggs/agg_type.ts:
```diff
@@ -224,6 +225,8 @@ export class AggType<
     return false;
   }
 
+  getResponseId: (agg: TAggConfig) => string;
```
Would be great to add a comment (document) this one explaining the current use-case as an example

Dosant replied:
This also should help with increased `Public APIs missing comments` stat

Dosant reviewed (approved):
LGTM

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.0](https://github.com/elastic/kibana/pull/120028)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/120028">](https://github.com/elastic/kibana/pull/120028) |

This backport PR will be merged automatically after passing CI.
